lib/engine/game/g_1858/step/track.rb

```diff
@@ -0,0 +1,183 @@
+              raise GameError, "The tile lay costs #{@game.format_currency(cost)} but " \
+                               "#{corporation.name} has #{@game.format_currency(corporation.cash)} and " \
+                               "#{president.name} has #{@game.format_currency(president.cash)}"
+            end
+            @game.log << "#{president.name} contributes #{@game.format_currency(shortfall)} " \
+                         "towards the cost of #{corporation.name}'s tile lay"
+            president.spend(shortfall, corporation)
+          end
+
+          # Override the default implementation to give a better log message.
+          # This also allows a more intuitive name to be used for the method for
+          # the player contributing money for the tile, instead of abusing
+          # try_take_loan.
+          def pay_tile_cost!(entity, tile, rotation, hex, spender, cost, _extra_cost)
+            subsidise_track(spender, cost)
+            spender.spend(cost, @game.bank) if cost.positive?
+
+            message = ''
+            message += "#{spender.name} spends #{@game.format_currency(cost)} and " if cost.positive?
+            message += "#{entity.name} " if cost.zero? || (entity != spender)
+            message += "lays tile ##{tile.name}"
+            message += " with rotation #{rotation} on #{hex.name}"
+            message += " (#{tile.location_name})" unless tile.location_name.to_s.empty?
+            @log << message
+          end
+
+          def hex_neighbors(entity, hex)
+            hexes_broad = @game.graph_broad.connected_hexes(entity)[hex]
+            hexes_metre = @game.graph_metre.connected_hexes(entity)[hex]
+            hexes = [hexes_broad, hexes_metre].compact.inject([], :|)
+            hexes if hexes.any?
+          end
+
+          def check_track_restrictions!(entity, old_tile, new_tile)
+            return if @game.loading || !entity.operator?
+
+            raise GameError, 'New track must override old one' if !@game.class::ALLOW_REMOVING_TOWNS &&
+                old_tile.city_towns.any? do |old_city|
+                  new_tile.city_towns.none? { |new_city| (old_city.exits - new_city.exits).empty? }
+                end
+
+            # Need to check twice whether this tile is OK, once using routes along
+            # broad/dual gauge track, and once allow metre/dual gauge.
+            if !valid_tile_lay?(entity, old_tile, new_tile, @game.graph_broad) &&
+                !valid_tile_lay?(entity, old_tile, new_tile, @game.graph_metre)
+              # # The check for private railways home hexes is needed in case a private
+              # # builds plain track that's not connected to a revenue centre, it will
+              # # not be classed as a connected path.
+              # XXX Is this still needed after the graphing changes? XXX
+              # && !(entity.minor? && entity.home_hex?(new_tile.hex))
+              raise GameError, 'Must use new track or change city value'
```

Comment:
Should this error message say something about being reachable by one of the track gauges?

Reply:
Good point. I have just reused the existing message for semi-restrictive upgrades, but I'll add something about track gauge.